src/compiler/checker.ts

```diff
@@ -23653,6 +23653,32 @@ namespace ts {
                 return type;
             }
 
+            function narrowTypeByPrivateIdentifierInInExpression(type: Type, expr: PrivateIdentifierInInExpression, assumeTrue: boolean): Type {
+                const target = getReferenceCandidate(expr.expression);
+                if (!isMatchingReference(reference, target)) {
+                    return type;
+                }
+
+                const privateId = expr.name;
+                const symbol = lookupSymbolForPrivateIdentifierDeclaration(privateId.escapedText, privateId);
+                if (symbol === undefined) {
+                    return type;
+                }
+                const classSymbol = symbol.parent!;
+                const classType = getTypeOfSymbol(classSymbol) as InterfaceType;
+                const classDecl = symbol.valueDeclaration;
+                Debug.assert(classDecl, "should always have a declaration");
+                let targetType: Type;
+                if (hasStaticModifier(classDecl)) {
+                    targetType = classType;
+                }
+                else {
+                    const classInstanceType = getDeclaredTypeOfSymbol(classSymbol);
+                    targetType = classInstanceType;
+                }

```

Comment:
```suggestion
                const targetType = hasStaticModifier(classDecl)
                    ? getTypeOfSymbol(classSymbol) as InterfaceType
                    : getDeclaredTypeOfSymbol(classSymbol);
```

(and `const classType` is no longer needed as well)